### PR TITLE
py-emodel-generalisation: add package

### PR DESCRIPTION
**Pull request by arnaudon**

### CODE REVIEW

**Comment by arnaudon** [comment]
requires https://github.com/BlueBrain/spack/pull/2149 to be merged first

**Comment by arnaudon** [comment]
I rebased on #2149 to check if it works

File: bluebrain/repo-bluebrain/packages/py-bluecellulab/package.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyBluecellulab(PythonPackage):
+    """Pythonic Blue Brain simulator access, former BGLibPy"""
+
+    homepage = "https://github.com/BlueBrain/BlueCelluLab"
+    git = "ssh://git@github.com:BlueBrain/BlueCelluLab.git"
```
Comment (matz-e) [inline bluebrain/repo-bluebrain/packages/py-bluecellulab/package.py]: This is already fixed in the other one

Comment (arnaudon) [reply bluebrain/repo-bluebrain/packages/py-bluecellulab/package.py]: Yes, I rebased on develop !

File: bluebrain/repo-bluebrain/packages/py-emodel-generalisation/package.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyEmodelGeneralisation(PythonPackage):
+    """Python library to generalise electrical models."""
+
+    homepage = "https://github.com/BlueBrain/emodel-generalisation"
+    git = "ssh://git@github.com:BlueBrain/emodel-generalisation.git"
```
Comment (matz-e) [inline bluebrain/repo-bluebrain/packages/py-emodel-generalisation/package.py]: do we have a PyPI URL?

Comment (arnaudon) [reply bluebrain/repo-bluebrain/packages/py-emodel-generalisation/package.py]: yes!

**Comment by arnaudon** [comment]
@matz-e , we need this for obp workflow, does it look ok to you? Thanks!

**Comment by eleftherioszisis** [comment]
I am wondering is it really essential to update all these dependencies, especially the py-pandas major? Could you keep the changes to a minimum so that it is also easier for you to add the package?

**Comment by arnaudon** [comment]
I'm not sure how to fix this one...

**Comment by matz-e** [comment]
> The following build jobs of the deployment for this PR failed:
> 
>     * [applications / science](https://bbpgitlab.epfl.ch/hpc/spack/-/jobs/1055692)

This is a known issue where different Python packages impose conflicting requirements on `setuptools`. I have started working on the next deployment iteration, in which Spack will treat these conflicts better.

**Comment by arnaudon** [comment]
Nice! Would you have tips on how to find out what package is making this conflict?

**Comment by matz-e** [comment]
> Nice! Would you have tips on how to find out what package is making this conflict?

I normally do a manual bisect on all dependencies. My guesses would be `numpy`, `scipy`, `pandas` or anything related.

**Comment by arnaudon** [comment]
its weird, I think i have the same deps as in other codes in term on numpy/pandas and friends I'm run out of ideas of things to try haha

**Comment by eleftherioszisis** [comment]
Seems that py-shap is if not the only, one of the culprits. @arnaudon is that even needed for emodel-generalisation? It seems you have many deps that are irrelevant to the emodel-generalisation pipeline. Maybe make them extras or soft dependencies?
https://bbpgitlab.epfl.ch/hpc/spack/-/jobs/1064901

**Comment by eleftherioszisis** [comment]
Seems however that the exact error is introduced by py-scikit-learn:
https://bbpgitlab.epfl.ch/hpc/spack/-/jobs/1065080
Caused by this:
```
depends_on("py-setuptools@:59", when="@1.0.2:", type="build")
```
which seems to have been replaced in the upstream with:
```
depends_on("py-setuptools@:59", when="@:1.2.1", type="build")
```

File: var/spack/repos/builtin/packages/py-scikit-learn/package.py [hunk 3]
```diff
@@ -60,7 +60,6 @@ class PyScikitLearn(PythonPackage):
 
     # pyproject.toml
     depends_on("py-setuptools", type="build")
-    depends_on("py-setuptools@:59", when="@1.0.2:", type="build")
```
Comment (eleftherioszisis) [inline var/spack/repos/builtin/packages/py-scikit-learn/package.py]: Can't do that. :P

Comment (arnaudon) [reply var/spack/repos/builtin/packages/py-scikit-learn/package.py]: shhht

**Comment by arnaudon** [comment]
@matz-e , let me know once you have merged the setuptool fix, I'll rebase. For the crash of bluecellulab, I thought this was fixed already in https://github.com/BlueBrain/spack/pull/2149 but apparently we're hitting this wall again. I didn't follow what was done about it..

**Comment by arnaudon** [comment]
@anilbey would you have a guess on why it fails with more recent bluecellulab version?

**Comment by arnaudon** [comment]
https://bbpgitlab.epfl.ch/hpc/module-testing/-/merge_requests/33 should fix it

**Comment by anilbey** [comment]
Judging from the error message I think it is trying to load the same mechanisms twice.

```
$ module load unstable py-bluecellulab neurodamus-neocortex
Autoloading python/3.10.8
Autoloading intel-oneapi-mkl/2021.4.0
Autoloading hpe-mpi/2.27.p1.hmpt
Autoloading python/3.10.8
Autoloading intel-oneapi-mkl/2021.4.0
Autoloading hpe-mpi/2.27.p1.hmpt
$ python scripts/cellulab_test.py
NEURON: The user defined name already exists: CaDynamics_DC0
 near line 0
 ^
        nrn_load_dll("/gpfs/bbp....")
	1 
Traceback (most recent call last):
  File "/gpfs/bbp.cscs.ch/ssd/gitlab_map_jobs/bbpcihpcproj12/J1128676/hpc/module-testing/scripts/cellulab_test.py", line 4, in <module>
    neuron.h.nrn_load_dll(os.environ["NRNMECH_LIB_PATH"])
RuntimeError: hocobj_call error: hoc_execerror: The user defined name already exists: CaDynamics_DC0
srun: error: r4i0n32: task 0: Exited with exit code 1
```

**Comment by arnaudon** [comment]
But I'm not sure why there is this loading in this test file, as bluecellulab loads it directly if the neurodamus module is available, no?

**Comment by anilbey** [comment]
> https://bbpgitlab.epfl.ch/hpc/module-testing/-/merge_requests/33 should fix it

Yes this issue was addressed previously, I think

**Comment by anilbey** [comment]
Yes you are right. Loading neurodamus sets `BLUECELLULAB_MOD_LIBRARY_PATH`, enabling bluecellulab to automatically read the mechanisms on that path.

```shell
[shell]~% module purge
Autoloading python/3.10.8
Autoloading intel-oneapi-mkl/2021.4.0
Autoloading hpe-mpi/2.27.p1.hmpt
[shell]~% echo $BLUECELLULAB_MOD_LIBRARY_PATH

[shell]~% module load unstable neurodamus-neocortex
Autoloading python/3.10.8
Autoloading intel-oneapi-mkl/2021.4.0
Autoloading hpe-mpi/2.27.p1.hmpt
[shell]~% echo $BLUECELLULAB_MOD_LIBRARY_PATH      
PATH_TO_IT/neurodamus-neocortex-1.13-2.17.0-2.8.1-ukvwcl/lib/libnrnmech_neocortex.so
[shell]~% 
```

**Review by matz-e — DISMISSED** [review]

**Comment by arnaudon** [comment]
so do you think I can remove the try/except and the load line in the test safely?

**Comment by arnaudon** [comment]
ok, @anilbey , forget it, its merged :)

**Comment by arnaudon** [comment]
I will merge the next version of this code, we have a little refinement in progress

**Comment by matz-e** [comment]
I'll try to remember to clean that test up later. I retriggered the testing part of the pipeline, should work now.

**Comment by arnaudon** [comment]
@matz-e I squeezed in a few more commits to avoid updating spack too soon, now its good to merge!

**Review by matz-e — APPROVED** [review]

**Pull request merged**